src/Utils/Utils.php

```diff
@@ -387,4 +386,40 @@ public static function filterSalesChannelDomainsWithoutHeadlessDomain(
             return !str_starts_with($domainEntity->getUrl(), Defaults::HEADLESS_SALES_CHANNEL_PREFIX);
         });
     }
+
+    /**
+     * Takes invalid URLs that contain special characters such as umlauts, or special UTF-8 characters and
+     * encodes them.
+     */
+    public static function getEncodedUrl(string $url): string
+    {
+        $parsedUrl = (array)parse_url($url);
+        $urlPath = explode('/', $parsedUrl['path']);
+        $encodedPath = [];
+        foreach ($urlPath as $path) {
+            $encodedPath[] = self::multiByteRawUrlEncode($path);
+        }
+
+        $parsedUrl['path'] = implode('/', $encodedPath);
+
+        return self::buildUrl($parsedUrl);
+    }
+
+    /**
+     * Flattens a given array. This method is similar to the JavaScript method "Array.prototype.flat()".
+     * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat
+     *
+     * @param array $array
+     *
+     * @return array
+     */
+    public static function flat(array $array): array
+    {
+        $flattened = [];
+        array_walk_recursive($array, static function ($a) use (&$flattened) {
+            $flattened[] = $a;
+        });
+
+        return $flattened;
+    }
```

Comment:
This should no longer be needed after refactoring.

Reply:
But do we still need it? Can't we get the data in a proper format already, so we don't need this? I imagine this could be very time consuming in case there are a lot of categories.

Reply:
We need the seo url of a product recursively along with its parents. So in the recursion we get an array of arrays. Which is then flattened out in the end outside the loop. So this is not called inside the recursion loop.